frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx

```diff
@@ -0,0 +1,143 @@
+import { useMemo } from "react";
+import { t } from "ttag";
+
+import { getColumnIcon } from "metabase/common/utils/columns";
+import {
+  getColumnGroupIcon,
+  getColumnGroupName,
+} from "metabase/common/utils/column-groups";
+
+import { Icon } from "metabase/core/components/Icon";
+import type { IconName } from "metabase/core/components/Icon";
+
+import * as Lib from "metabase-lib";
+
+import { StyledAccordionList } from "../FilterPicker.styled";
+
+export interface FilterColumnPickerProps {
+  query: Lib.Query;
+  stageIndex: number;
+  onColumnSelect: (column: Lib.ColumnMetadata) => void;
+  onSegmentSelect: (segment: Lib.SegmentMetadata) => void;
+  onExpressionSelect: () => void;
+}
+
+type Section = {
+  key?: string;
+  name: string;
+  items: (Lib.ColumnMetadata | Lib.SegmentMetadata)[];
+  icon?: IconName;
+};
+
+const CUSTOM_EXPRESSION_SECTION: Section = {
+  key: "custom-expression",
+  name: t`Custom Expression`,
+  items: [],
+  icon: "filter",
+};
+
+export type ColumnListItem = Lib.ColumnDisplayInfo & {
+  column: Lib.ColumnMetadata;
+};
+
+export type SegmentListItem = Lib.SegmentDisplayInfo & {
+  segment: Lib.SegmentMetadata;
+};
+
+const isSegmentListItem = (
+  item: ColumnListItem | SegmentListItem,
+): item is SegmentListItem => {
+  return (item as SegmentListItem).segment !== undefined;
+};
+
+/**
+ * Select a column, segment, or custom expression upon which to filter
+ * Filter ColumnOrSegmentOrCustomExpressionPicker was too long of a name
+ */
+export function FilterColumnPicker({
+  query,
+  stageIndex,
+  onColumnSelect,
+  onSegmentSelect,
+  onExpressionSelect,
+}: FilterColumnPickerProps) {
+  const sections = useMemo(() => {
+    const columns = Lib.filterableColumns(query, stageIndex);
+    const columnGroups = Lib.groupColumns(columns);
+
+    const sections = columnGroups.map(group => {
+      const groupInfo = Lib.displayInfo(query, stageIndex, group);
+
+      const columnItems = Lib.getColumnsFromColumnGroup(group).map(column => ({
+        ...Lib.displayInfo(query, stageIndex, column),
+        column,
+      }));
+
+      const includeSegments = !groupInfo.isFromJoin;
```

Comment:
```suggestion
      const includeSegments = groupInfo.isSourceTable;
```

With `!groupInfo.isFromJoin` we'd also include segments in implicitly joinable table groups